rooms: move height vars and functions to TRX

lahm86 opened the pull request:
Resolves #2541.

#### Checklist

- [x] I have read the [coding conventions](https://github.com/LostArtefacts/TRX/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] I have added a changelog entry about what my pull request accomplishes, or it is an internal change
- [x] I have added a readme entry about my new feature or OG bug fix, or it is a different change

#### Description

This primarily replaces `g_HeightType` and `g_GF_NoFloor` with function calls in TRX, but also moves `Room_GetHeight` and `Room_GetCeiling` together as well for consistency. There remains some height specific logic in the separate modules, for example altering floor height, but that's linked with #2036 so can be handled separately there.

It seemed sensible to offer `disable_floor` in TR1 as part of this as it's quite a straight-forward feature.

For testing, we can check floor and ceiling heights are OK, including slopes, plus the abyss in Floating Islands. You can also try an abyss in TR1; one I tried was setting the value to 4096 in Great Pyramid and then dropping off the starting ledge. It doesn't really make sense in OG environments though, so it's definitely more of a custom level feature.

lahm86 committed:
rooms: move no floor marker to TRX

This moves TR2's no floor indicator to TRX's room module.

lahm86 committed:
tr1/game_flow: add support for no floor

This adds support for no floor in TR1, in the same style as TR2's
Floating Islands.

Resolves #2541.

lahm86 committed:
rooms: move floor height function to TRX

This moves Room_GetHeight to TRX, which in turn allows us to eliminate
g_HeightType.

lahm86 committed:
rooms: move ceiling height function to TRX

This moves Room_GetCeiling to TRX.

rr- commented on src/tr2/game/game_flow/sequencer_events.c:
```diff
@@ -200,7 +200,7 @@ static DECLARE_GF_EVENT_HANDLER(M_HandleDisableFloor)
     GF_COMMAND gf_cmd = { .action = GF_NOOP };
     if (seq_ctx != GFSC_STORY) {
-        g_GF_NoFloor = (int16_t)(intptr_t)event->data;
+        Room_SetNoFloorHeight((int16_t)(intptr_t)event->data);
```
Maybe `SetAbyssHeight`? or `SetVoidHeight`. I'm also perfectly fine with the current naming.

lahm86 replied:
Thanks, I think `Abyss` is a good choice; `NoFloor` was a bit awkward.

rr- reviewed (approved):
LGTM and thank you.

lahm86 committed:
rooms: rename no floor internals

This renames the no floor internal concept to abyss.

aredfan reviewed (approved):
Overall LGTM.

A potential issue I noticed is when I use the `disable_floor` in Great Pyramid, there's also invisible collision around the top platform that Lara will bump into if she flies around.  Is that to be expected?

lahm86 committed:
rooms: use calculated max abyss height

Rather than hardcoding the maximum abyss height to 16384, we calculate
it based on the OG difference between this and Floating Islands abyss
height. This avoids potential collision issues near the top of the
level.

lahm86 commented:
Thanks @aredfan. That's a good point, and could cause issues in TR2 customs as well; the hard-coding of 16384 as the absolute abyss maximum causes this, as the height at the top of the platform in GP causes things to overflow (note there is other collision here just beyond the scion platform and near the sloped ceiling, but that's OG). I've adjusted it now so it calculates the absolute floor, based on the difference between 16384 and OG floater abyss value. It'd be worth flying around near the roof in Floater to see if everything looks ok with this, and see how GP looks too. Thanks!

aredfan reviewed (approved):
Thank you for explaining.  The issue is fixed, and the Floater and GP levels both looks good. 👍

rr- commented on src/libtrx/game/rooms/common.c:
```diff
@@ -511,12 +512,14 @@ SECTOR *Room_GetSkySector(
 {
-    m_AbyssHeight = height;
+    m_AbyssMinHeight = height;
+    m_AbyssMaxHeight = height == 0 ? 0 : m_AbyssMinHeight + 30 * STEP_L;
```
Hmm, what does `30 * STEP_L` signify? Asking cause it doesn't match 16384 (16 tiles).
Also would be good to explain what is the min and max role in the calculations, as it's become quite confusing to me. I always thought it's a simple plane.

lahm86 replied:
My bad, I got my numbers muddled. I've set it now to `26 * STEP_L`, so this is the difference between 16384 and Floater abyss height of 9728. This allows time when Lara is falling between setting her HP to 0 and the inventory showing, so that we don't hear Lara hit the floor. Although I've noticed that in OG, you can _just_ hear the crunch SFX begin as the inventory opens, but we can perhaps address that separately.
We also clamp to just under the max height; in TombEditor in theory you could build a room this low.
I've added comments, hopefully they clear things up.
